File: src/main/java/mainPackage/CustomTask.java
package mainPackage;

import javafx.application.Platform;
import javafx.concurrent.Task;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * Created by jacobmenke on 4/17/17.
 */
public class CustomTask<T> extends Task<String> {
    Runnable r;
    MainController mainController;
    ExecutorService executorService;
    Future future;
    boolean searchingTaskRunning = false;

    public CustomTask(MainController mainController, Runnable r, boolean b) {
        this.mainController = mainController;
        this.r = r;
        searchingTaskRunning = b;
    }

    public CustomTask(MainController mainController) {
        this.mainController = mainController;
    }

    public CustomTask() {
    }

    public CustomTask(MainController mainController, Runnable r) {
        this.mainController = mainController;
        this.r = r;
    }

    public Future getFuture() {
        return future;
    }

    public Runnable getRunnable() {
        return r;
    }

    public void setRunnable(Runnable r) {
        this.r = r;
    }

    public ExecutorService getExecutorService() {
        return executorService;
    }

    @Override
    public void updateMessage(String message) {
        super.updateMessage(message);
    }

    @Override
    protected void updateProgress(double workDone, double max) {
        super.updateProgress(workDone, max);
    }

    @Override
    protected void updateTitle(String title) {
        super.updateTitle(title);
    }

    @Override
    protected void updateValue(String value) {
        super.updateValue(value);
    }

    @Override
    protected String call() throws Exception {

        executorService = Executors.newSingleThreadExecutor();
        future = executorService.submit(r);
        //System.out.println("called and submitting to executor");

        try {
            //System.out.println("waiting on future");
            future.get();
            //System.out.println("future done");

        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        } finally {
            if (!searchingTaskRunning) {
                Platform.runLater(() -> {
                    //      Utilities.removeFromView(mainController.thinkingIndicator);
                    Utilities.removeFromView(mainController.sphere);
                    mainController.timeline.stop();
                });
            }
        }

        executorService.shutdownNow();

        System.gc();

        return "Completed";
    }
}

File: src/main/kotlin/mainPackage/CustomTask.kt
package mainPackage

import javafx.application.Platform
import javafx.concurrent.Task
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

open class CustomTask<T> : Task<String> {
    var r: Runnable? = null
    var mainController: MainController? = null
    var executorService: ExecutorService? = null
    var future: Future<*>? = null
    var searchingTaskRunning = false

    constructor(mainController: MainController, r: Runnable, b: Boolean) {
        this.mainController = mainController
        this.r = r
        this.searchingTaskRunning = b
    }

    constructor(mainController: MainController) {
        this.mainController = mainController
    }

    constructor()

    constructor(mainController: MainController, r: Runnable) {
        this.mainController = mainController
        this.r = r
    }

    fun getRunnable(): Runnable? = r
    fun setRunnable(r: Runnable) { this.r = r }

    public override fun updateMessage(message: String?) { super.updateMessage(message) }
    public override fun updateProgress(workDone: Double, max: Double) { super.updateProgress(workDone, max) }
    public override fun updateTitle(title: String?) { super.updateTitle(title) }
    public override fun updateValue(value: String?) { super.updateValue(value) }

    override fun call(): String {
        executorService = Executors.newSingleThreadExecutor()
        future = executorService!!.submit(r)

        try {
            future!!.get()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: java.util.concurrent.ExecutionException) {
            e.printStackTrace()
        } finally {
            executorService!!.shutdownNow()
            executorService!!.awaitTermination(2, TimeUnit.SECONDS)
            if (!searchingTaskRunning) {
                Platform.runLater {
                    Utilities.removeFromView(mainController!!.sphere)
                    mainController!!.timeline?.stop()
                }
            }
        }
        return "Completed"
    }
}
